web/modules/custom/joinup_core/joinup_core.post_update.php

```diff
@@ -1136,3 +1136,81 @@ function joinup_core_post_update_delete_orphaned_solutions() {
     $solution->delete();
   }
 }
+
+/**
+ * Fix all needed timestamps and created dates for content.
+ */
+function joinup_core_post_update_0001_set_migrated_content_timestamp(&$sandbox) {
+  $connection = \Drupal::database();
+
+  // Set the oldest revision timestamp to the created date for content.
+  // The created time was retained throughout the migration but the revision
+  // timestamp was set anew during the migration.
+  $query = <<<QUERY
+UPDATE {node_revision} nr
+INNER JOIN (
+    SELECT nid, MIN(created) AS created_timestamp
+    FROM node_field_revision
+    GROUP BY nid, vid
+    ORDER BY vid ASC
+  ) nfr
+  ON nr.nid = nfr.nid
+SET nr.revision_timestamp = nfr.created_timestamp
+WHERE nr.nid < 700000;
+QUERY;
+  $connection->query($query)->execute();
+
+  // Set the created timestamp all content revisions to the earliest revision
+  // timestamp of that node. This will undo our workaround change the created
+  // time upon initial publication.
+  $query = <<<QUERY
+UPDATE {node_field_revision} nfr
+ INNER JOIN (
+    SELECT nid, MIN(revision_timestamp) as revision_timestamp
+    FROM node_revision
+    GROUP BY nid
+) nr ON nfr.nid = nr.nid
+SET nfr.created = nr.revision_timestamp
+WHERE nfr.nid = nr.nid AND nfr.nid >= 700000
+QUERY;
+  $connection->query($query)->execute();
+
+  // Update also the node_field_data table.
+  $query = <<<QUERY
+UPDATE {node_field_data} nfd
+  INNER JOIN (
+    SELECT nid, vid, created
+    FROM node_field_revision
+) nfr ON nfd.vid = nfr.vid
+SET nfd.created = nfr.created
+WHERE nfd.vid = nfr.vid
+QUERY;
+  $connection->query($query)->execute();
+
+  // Update all publication dates according to the minimum possible revision
+  // timestamp.
+  $query = <<<QUERY
+UPDATE {node_field_revision} r, (
+  SELECT
+    subnfr.nid,
+    MIN(subnfr.vid) as vid,
+    MIN(subnr.revision_timestamp) as revision_timestamp
+  FROM {node_field_revision} subnfr
+  INNER JOIN {node_revision} subnr ON subnfr.vid = subnr.vid
+  WHERE status = 1
+  GROUP BY nid
+  ORDER BY vid
+) s
+SET r.published_at = s.revision_timestamp
+WHERE r.nid = s.nid AND r.vid >= s.vid;
```

Comment:
I'm not sure I understand `AND r.vid >= s.vid`.

Reply:
Consider this
| nid | vid | status | published_at |
| 1    | 1   | 0         | DEFAULT      |
| 1    | 2   | 1         | 1                    |
| 1    | 3   | 0         | 1                    |
| 1    | 4   | 1         | 1                    |

We had a case where in a node, an intermediate unpublished revision (vid 3 in the example above), had the default value instead of the publication timestamp.
The programmed functionality of the publication date is that once the entity is published, every revision takes the publication date regardless of the status as it _has_ been published in the past.
The `r.vid >= s.vid` means that every revision after the one that was initially published, should receive the publication date.